xrmocap/transform/keypoints3d/optim/rm_duplicate.py

```diff
@@ -0,0 +1,143 @@
+
+    def optimize_keypoints3d(self, keypoints3d: Keypoints,
+                             **kwargs: dict) -> Keypoints:
+        """Forward function for RemoveDuplicate optimizor.
+
+        Args:
+            keypoints3d (Keypoints):
+                Keypoints3d to be optimized.
+
+        Returns:
+            Keypoints: Optimized keypoints3d.
+        """
+        # get kps3d array
+        keypoints3d_optim = keypoints3d.clone()
+        kps3d = keypoints3d_optim.get_keypoints()
+        n_frame, n_max_person, n_kps, _ = kps3d.shape
+
+        kps3d_optim = np.full((n_frame, n_max_person, n_kps, 4), np.nan)
+
+        for frame_idx in range(n_frame):
+            kps3d_frame = kps3d[frame_idx, ...]
+            kps3d_frame = kps3d_frame[~np.isnan(kps3d_frame[:, 0, 0])]
+
+            # calculate distance and remove duplicate
+            dist = self.get_kps3d_dist(kps3d_frame)
+
+            remove_idxs = []
+            keep_idxs = []
+            for person_idx, per_person_dist in enumerate(dist):
+                if person_idx not in remove_idxs \
+                        and person_idx not in keep_idxs:
+                    to_remove = np.where(
+                        per_person_dist[:] < self.threshold)[0]  # index
+
+                    if self.keep == 'by_conf':
+                        # keep the highest confidence
+                        keep_idx = to_remove[np.argmax(kps3d_frame[to_remove,
+                                                                   0, -1])]
+
+                    elif self.keep == 'by_index':
+                        # keep the lowest index
+                        keep_idx = person_idx
+
+                    to_keep = np.array([keep_idx])
+                    to_remove = np.setdiff1d(to_remove, to_keep)
+
+                    remove_idxs.extend(list(to_remove))
+                    keep_idxs.extend(list(to_keep))
+
+            # identity tracking
+            if self.identity_tracking is not None:
+                curr_kps3d = kps3d_frame[keep_idxs, ...]
+                frame_identities = self.identity_tracking.query(curr_kps3d)
+
+                # save to Kps3d
+                kps3d_optim[frame_idx, frame_identities,
+                            ...] = kps3d_frame[keep_idxs, ...]
+            else:
+                # save to Kps3d
+                n_optim_person = len(keep_idxs)
+                kps3d_optim[frame_idx, :n_optim_person,
+                            ...] = kps3d_frame[keep_idxs, ...]
+
+        keypoints3d_optim.set_keypoints(kps3d_optim)
+        keypoints3d_optim.set_mask(kps3d_optim[..., -1] > 0)
+
+        return keypoints3d_optim
+
+    def get_kps3d_dist(self,
+                       kps3d: Union[torch.Tensor, np.ndarray],
+                       p: int = 2) -> np.ndarray:
+        """Calculate the distance between each set of keypoints3d.
+
+        Args:
+            kps3d (Union[torch.Tensor, np.ndarray]):
+                keypoints3d of the current frame
```

Comment:
Doc the shape of kps3d

Reply:
Fixed in [c2ba151](https://github.com/openxrlab/xrmocap/pull/46/commits/c2ba1517e1c5c68dd276bf21e002f6e4ef4df715).